File: src/ie/states/pa.ts
import { removeNonDigits } from "../../shared/normalize.js";
import { weightedSum, mod11CheckDigit } from "../../shared/mod11.js";

// Verified against the official SEFAZ-PA "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_PA.html), including its two
// worked examples (15999999-5, 75000002-3). Format: 8 digits (always
// starting with 15, 75, 76, 77, 78 or 79) + 1 check digit.
const WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2];
const VALID_PREFIXES = new Set(["15", "75", "76", "77", "78", "79"]);

function normalize(value: string): string {
  return removeNonDigits(value);
}

function isValid(value: string): boolean {
  if (!/^[\d\-\s]+$/.test(value)) {
    return false;
  }

  const ie = normalize(value);

  if (ie.length !== 9 || !VALID_PREFIXES.has(ie.slice(0, 2))) {
    return false;
  }

  return mod11CheckDigit(weightedSum(ie, WEIGHTS)) === Number(ie[8]);
}

function format(value: string): string {
  const ie = normalize(value);

  if (ie.length !== 9) {
    return value;
  }

  return ie.replace(/^(\d{8})(\d)$/, "$1-$2");
}

export const PA = {
  isValid,
  normalize,
  format,
};

File: java/src/main/java/io/github/matheuslm7/brvalidator/ie/Pa.java
package io.github.matheuslm7.brvalidator.ie;

import io.github.matheuslm7.brvalidator.internal.Shared;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * PA: verified against the official SEFAZ-PA "Roteiro de Crítica da Inscrição Estadual"
 * (sintegra.gov.br/Cad_Estados/cad_PA.html), including its two worked examples (15999999-5,
 * 75000002-3). Format: 8 digits (always starting with 15, 75, 76, 77, 78 or 79) + 1 check digit.
 */
final class Pa {

  private static final Pattern ALLOWED_CHARS = Pattern.compile("^[\\d\\-\\s]+$");
  private static final int[] WEIGHTS = {9, 8, 7, 6, 5, 4, 3, 2};
  private static final Set<String> VALID_PREFIXES =
      Set.of("15", "75", "76", "77", "78", "79");

  private Pa() {}

  static String normalize(String value) {
    return Shared.removeNonDigits(value);
  }

  static boolean isValid(String value) {
    if (!ALLOWED_CHARS.matcher(value).matches()) {
      return false;
    }

    String v = normalize(value);

    if (v.length() != 9 || !VALID_PREFIXES.contains(v.substring(0, 2))) {
      return false;
    }

    return Mod11.mod11CheckDigit(Mod11.weightedSum(v, WEIGHTS)) == (v.charAt(8) - '0');
  }

  static String format(String value) {
    String v = normalize(value);

    if (v.length() != 9) {
      return value;
    }

    return v.substring(0, 8) + "-" + v.substring(8, 9);
  }
}
